=str #18556 add delay combinator

agolubev opened the pull request:
Ref #18556 
Seems like trivial. 
Also removed redundant code like `override def onUpstreamFinish(): Unit = completeStage()`

agolubev committed:
=str #18556 add delay combinator

ktoso commented:
OK TO TEST

drewhk commented:
This is not necessarily what I would consider to be delay semantics, this is more like a minimum-time-between operation. To demonstrate what I mean:

 - assume original stream emits elements A and B at times 1 and 5
 - delay is 10
 - this op will emit A at 11 and B at 21 (ignoring element passing costs), since it schedules the 10 seconds delay for B *after* A has been emitted, and not when B was actually received.

Now the delay semantics that require an element to be scheduled to be around T + Delta (assuming backpressure does not interfere) can be only obtained by an unbounded internal buffer. This is quite costly, and backpressure might mess up the accurate delaying anyway, nevertheless, this is the more intuitive delay operation.

I am not sure what are the right semantics to aim for here. @rkuhn @viktorklang any opinions?

rkuhn commented:
I agree with @drewhk, the delay needs to start ticking upon reception of each element. The buffer of course needs to be bounded, which then means that the user needs to select how to react to when it runs full. The choices are

* emit early
* drop
* fail
* back-pressure

The last one may or may not make sense, depending on the use-case, but it does have clearly defined semantics (and it also degenerates into a rate limiter plus delay).

drewhk commented:
I am not sure we need to provide more options than early and backpressure, since a buffer can be always attached in front of it. In fact this might mean that a simple buffer of one might be just enough.

rkuhn commented:
good point

agolubev commented:
Make sense, I was looking here: http://reactivex.io/documentation/operators/delay.html
But current ticket is about minimum rate for emitting elements. 
Ok, I'll will rework this

drewhk commented:
> But current ticket is about minimum rate for emitting elements. 

The ticket is about delaying... What do I miss here?

> Make sense, I was looking here: http://reactivex.io/documentation/operators/delay.html

Interestinly enough, their diagram illustrates what me and @rkuhn proposed, while their description directly contradicts that: "The Delay operator modifies its source Observable by pausing for a particular increment of time (that you specify) before emitting each of the source Observable’s items.". If their diagram would depict what they just said then every marble would be at least delay space apart, which they aren't.

drewhk commented:
Ah, because in Rx they don't need to request for next. But that means they will need many timers

agolubev commented:
@drewhk, actually the question is should delay starting point be when:
1) last element pushed (when A sends downstream in you example)
2) when received current element to push (when we get B)
The second one is what PR is doing now.

drewhk commented:
@agolubev we proposed 2, but with the difference that you need to request elements in advance, othewise you would never learn their true arrival time.

drewhk commented:
To give a small test case:

```
Source(1 to 10).delay(10.seconds).completionTimeout(11.seconds)
```

should not throw, since every element is delayed compared to its true arrival (i.e. just as if the delay operation would not be there). With this PR it would take 100 seconds to finish because there would be 10 seconds delay between *each* element.

agolubev commented:
Ok, got it. 
>In fact this might mean that a simple internal buffer of one might be just enough.

Per my understanding we need to store elements during delay time somewhere. So not sure how to operate external (upstream?) buffer in this case. Are there any example in existing code base I can refer to?

agolubev commented:
I'm thinking about having buffer functionality (with overflow strategy) separated. And then use it in Buffer detached state and in delay GraphStage

drewhk commented:
Unfortunately I don't think a separate buffer will work (I was wrong with that), so probably the best is to have a simple Array that can hold up to N elements with their arrival timestamp while the next element is being delayed. Then use the drop strategy to decide what to do (as @rkuhn proposed).

agolubev committed:
+str #18556 add delay combinator

agolubev commented:
@drewhk, please take a look when you have time. 
- I added strategies for delay
- now like API more, still not sure if need to add buffer size as a parameter
- added trait to mark OverflowStrategies as Delay related, this change failed few tests with OverflowSrategy  serialization check failed. Also this change caused warning (fixed with suppressing this) for match expression with OverflowStrategy.

agolubev commented:
Build failed as failed Auto-merging akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala

drewhk commented on akka-stream-testkit/src/main/scala/akka/stream/testkit/StreamTestKit.scala:
```diff
@@ -266,6 +266,16 @@ object TestSubscriber {
     /**
      * Fluent DSL
      *
+     * Expect a stream element during specified time, then timeout.
```
or instead of then

drewhk commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowDelaySpec.scala:
```diff
@@ -0,0 +1,118 @@
+
+  implicit val materializer = ActorMaterializer()
+
+  "A Delay" must {
```
add also a test of `Source(1 to 10).initialDelay()` which will delay the initial element of the source (but not others). Then you can see if your `delay` combinator properly adds time to that initial delay.

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
this case is a bug, so this should throw an exception indicating that this should not happen.

drewhk commented on akka-stream/src/main/scala/akka/stream/OverflowStrategy.scala:
```diff
@@ -6,42 +6,45 @@ package akka.stream
   /**
    * INTERNAL API
    */
-  private[akka] final case object DropHead extends OverflowStrategy
```
why remove the final?

agolubev replied:
IntelliJ told it's redundant, it seemed for me logical

hepin1989 replied:
@agolubev I asked the question once and akka team said that they mark it final for internal thing,

agolubev replied:
Yeah, will look deeply into this on weekend.

drewhk replied:
no I didn't see it is a case object. I thought it is a case class.

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
why do we need to schedule a timer here?

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
it should basically alway pull to the buffer, except when it is full && Backpressure.

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
shouldn't a timer be fired if the buffer was empty just before the element arrived?

agolubev commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
yeah - got it - so will have backpressure only on full buffer

agolubev commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
refactor this a little. now it's scheduing timer when buffer is not full

agolubev commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
not sure I'm following you. Timer is pushing elements and schedule next time if buffer is not empty. So not sure why need to call onTimer before element arrived.

agolubev committed:
+str #18556 add delay combinator

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
I would give `nextElementWaitTime` a small threshold since it is not worth to fire a timer of 10ms resolution when the waiting time is less than 10ms.

drewhk commented:
LGTM. Please rebase.

viktorklang commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
The problem is that you can't make any assumption on the Scheduler implementation, unless you want to introspect the config for the LARS and check its tick-duration?

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
I still find it overkill to schedule timers for such a short delay. In worst case, you can schedule a timer for every message even if they arrived basically subsequently. We can accept an explicit resolution parameter, but the default of 10ms is not arbitrary since this is what you can reasonably expect from most OSs. Of course multimedia timers might be more precise, but that is a bit out of scope.

viktorklang commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
Where can I find the reference for the 10ms reasonability? I thought high-precision times could have sub-ms precision?

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
> For example, for Windows running on an x86 processor, the default interval between system clock ticks is typically about 15 milliseconds,

(https://msdn.microsoft.com/en-us/library/windows/hardware/dn265247(v=vs.85).aspx)

To get more resolution there are other Timer versions for higher resolution but you cannot rely on this on the JVM. (at least for a while)

viktorklang commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
Speaking of which, this code in this PR ought to use nanoTime over currentTimeMillis for resolution purposes.

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
Also: 
> Practical clocks tick at a finite rate, with rates of 100 hertz and 1000 hertz being common. The inverse of this tick rate is the clock resolution, also called the clock granularity, which in either case is expressed as a time duration, being 10 milliseconds and 1 millisecond respectively for these common rates

(timer_settimer manpage: http://linux.die.net/man/3/timer_settime)

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
yes nanoTime is a good idea.

viktorklang commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
I think a good compromise is to cap the precision of delay to 10ms and document that properly, and then we'll see if anyone complains.

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
The nice thing is that delay will not accumulate delay errors though, since eventually the error gets larger than the 10ms threshold, which will trigger a timer that will just automatically compensate previous errors (since we know the exact arrival time of elements).

viktorklang commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
Hehe, it will *converge* on correctness if the stream is long enough ;)

agolubev committed:
+str #18556 add delay combinator

agolubev commented:
@drewhk, ready for review
build failed with CONFLICT (content): Merge conflict in akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala
